### ISSUE TITLE
[xray] Allowing Fractional use of the GPU with repeat > 1

### ISSUE THREAD

**Issue by makdoudN** [ray-project/ray#2593]
Thank you for open sourcing this project.

### System information
- **Linux Ubuntu 16.04**:
- **binary**:
- **0.5**:
- **3.5/3.6 (anaconda)**

### Describe the problem

Is it possible to use a fraction on my GPU with the option repeat in order to use as much as possible my GPU.

The context is in RL depending of the algorithm and / or the environment the bottleneck could prevent the use of N% of the GPU. It is interesting in this context to take advantage of the bottleneck to allow the repetition of the ray.tune.Experiment by allowing to parallelize training with partial GPU usage.

I know #402 tracks fractional use of the GPU is being implemented in the new backend.

Do you think this feature is interesting and possible to be implemented with ray after #402 ?

Best regards.
Makdoud

### Source code / logs

When I try to use fractional use of the gpu (gpus=0.2) with repeat=2

I will post the exact command in order to reproduce the following traceback in few hours.

Created LogSyncer for /home/nizam/ray_results/Learning-by-Interact/main_1_2018-07-19_10-30-54xckffkd6 ->
Created LogSyncer for /home/nizam/ray_results/Learning-by-Interact/main_2_2018-07-19_10-30-54myss4all ->
/ray/src/local_scheduler/local_scheduler.cc:451 Check failed: state->available_gpus.size() >= resource_quantity
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler[0x442535]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_ZN3ray8internal7CerrLogD1Ev+0x8e)[0x44f17e]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z17acquire_resourcesP19LocalSchedulerStateP20LocalSchedulerClientRKSt13unordered_mapISsdSt4hashISsESt8equal_toISsESaISt4pairIKSsdEEE+0x36b)[0x44817b]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z21assign_task_to_workerP19LocalSchedulerStateR17TaskExecutionSpecP20LocalSchedulerClient+0x60)[0x44b040]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z14dispatch_tasksP19LocalSchedulerStateP24SchedulingAlgorithmState+0xee)[0x4542de]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z18dispatch_all_tasksP19LocalSchedulerStateP24SchedulingAlgorithmState+0x13)[0x455fb3]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z23handle_worker_availableP19LocalSchedulerStateP24SchedulingAlgorithmStateP20LocalSchedulerClient+0x1da)[0x4565ea]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(_Z15process_messageP11aeEventLoopiPvi+0xa4e)[0x44dc6e]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(aeProcessEvents+0x227)[0x482717]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(aeMain+0x2b)[0x4829bb]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler(main+0x5a7)[0x442eb7]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7f55706cc830]
/home/nizam/anaconda3/envs/py36/lib/python3.6/site-packages/ray/local_scheduler/../core/src/local_scheduler/local_scheduler[0x444e91]

**Comment by richardliaw** [comment]
Sorry for the delayed response! This somehow slipped through the cracks.

@robertnishihara, did you end up implementing fractional resource support?

If not, @makdoudN perhaps you can use the workaround listed in the issue you mentioned.

**Comment by robertnishihara** [comment]
@richardliaw yes, fractional GPU support is implemented and should be working in the new backend.

@makdoudN can you clarify what you mean by "repeat > 1"?

**Comment by richardliaw** [comment]
IIUC "repeat" is the Tune configuration parameter that specifies the number of samples.

**Comment by makdoudN** [comment]
Using: [pbt_tune_cifar10_with_keras.txt](https://github.com/ray-project/ray/files/2289706/pbt_tune_cifar10_with_keras.txt).
A simple modification of pbt_tune_cifar10_with_keras.txt with pbt disable, raylet true, a batch of 2 and a network modification in order to consume less that 50% of my gpu. It crashes with gpu=.4 but works with gpu=.6 with repeat > 0.

The traceback is the following:
 ```
Process STDOUT and STDERR is being redirected to /tmp/raylogs/.                                                                                                                                                                               
Waiting for redis server at 127.0.0.1:51694 to respond...                                                                                                                                                                                     
Waiting for redis server at 127.0.0.1:12417 to respond...                                                                                                                                                                                     
/ray/src/ray/raylet/main.cc:35: Starting raylet with static resource configuration: {GPU,1.000000}, {CPU,8.000000},                                                                                                                           
/ray/src/ray/raylet/main.cc:66: Starting object manager with configuration:                                                                                                                                                                   
max_sends = 2                                                                                                                                                                                                                                 
max_receives = 2                                                                                                                                                                                                                              
object_chunk_size = 1000000                                                                                                                                                                                                                   
/ray/src/ray/raylet/main.cc:76: Initializing GCS client bcf600803af011a80e3a5345f8e48d21e6db7449                                                                                                                                              
                                                                                                                                                                                                                                              
======================================================================                                                                                                                                                                        
View the web UI at http://localhost:8892/notebooks/ray_ui89793.ipynb?token=065e0c111ea81fce04f84d5d5bb7da6eb290f1e4705b51ba                                                                                                                   
======================================================================                                                                                                                                                                        
                                                                                                                                                                                                                                              
== Status ==                                                                                                                                                                                                                                  
Using FIFO scheduling algorithm.                                                                                                                                                                                                              
Result logdir: /home/nizam/ray_results/pbt_cifar10                                                                                                                                                                                            
PENDING trials:                                                                                                                                                                                                                               
 - train_cifar10_0_decay=1e-06,dropout=0.25,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_10_decay=1e-07,dropout=0.25,lr=1e-05:  PENDING                                                                                                                                                                               
 - train_cifar10_11_decay=1e-07,dropout=0.5,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_12_decay=1e-06,dropout=0.25,lr=0.0001: PENDING                                                                                                                                                                               
 - train_cifar10_13_decay=1e-06,dropout=0.5,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_14_decay=1e-07,dropout=0.25,lr=1e-05:  PENDING                                                                                                                                                                               
 - train_cifar10_15_decay=1e-07,dropout=0.5,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_1_decay=1e-06,dropout=0.5,lr=0.0001:   PENDING                                                                                                                                                                               
 - train_cifar10_2_decay=1e-07,dropout=0.25,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_3_decay=1e-07,dropout=0.5,lr=1e-05:    PENDING                                                                                                                                                                               
 - train_cifar10_4_decay=1e-06,dropout=0.25,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_5_decay=1e-06,dropout=0.5,lr=0.0001:   PENDING                                                                                                                                                                               
 - train_cifar10_6_decay=1e-07,dropout=0.25,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_7_decay=1e-07,dropout=0.5,lr=1e-05:    PENDING                                                                                                                                                                               
 - train_cifar10_8_decay=1e-06,dropout=0.25,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_9_decay=1e-06,dropout=0.5,lr=0.0001:   PENDING                                                                                                                                                                               
                                                                   
Created LogSyncer for /home/nizam/ray_results/pbt_cifar10/train_cifar10_0_decay=1e-06,dropout=0.25,lr=0.0001_2018-08-15_10-47-427fbe82yn ->                                                                                                   
== Status ==                                                                                                                                                                                                                                  
Using FIFO scheduling algorithm.                                                                                                                                                                                                              
Resources requested: 1/8 CPUs, 0.4/1 GPUs                                                                                                                                                                                                     
Result logdir: /home/nizam/ray_results/pbt_cifar10                                                                                                                                                                                            
PENDING trials:                                                                                                                                                                                                                               
 - train_cifar10_10_decay=1e-07,dropout=0.25,lr=1e-05:  PENDING                                                                                                                                                                               
 - train_cifar10_11_decay=1e-07,dropout=0.5,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_12_decay=1e-06,dropout=0.25,lr=0.0001: PENDING                                                                                                                                                                               
 - train_cifar10_13_decay=1e-06,dropout=0.5,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_14_decay=1e-07,dropout=0.25,lr=1e-05:  PENDING                                                                                                                                                                               
 - train_cifar10_15_decay=1e-07,dropout=0.5,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_1_decay=1e-06,dropout=0.5,lr=0.0001:   PENDING                                                                                                                                                                               
 - train_cifar10_2_decay=1e-07,dropout=0.25,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_3_decay=1e-07,dropout=0.5,lr=1e-05:    PENDING                                                                                                                                                                               
 - train_cifar10_4_decay=1e-06,dropout=0.25,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_5_decay=1e-06,dropout=0.5,lr=0.0001:   PENDING                                                                                                                                                                               
 - train_cifar10_6_decay=1e-07,dropout=0.25,lr=1e-05:   PENDING                                                                                                                                                                               
 - train_cifar10_7_decay=1e-07,dropout=0.5,lr=1e-05:    PENDING                                                                                                                                                                               
 - train_cifar10_8_decay=1e-06,dropout=0.25,lr=0.0001:  PENDING                                                                                                                                                                               
 - train_cifar10_9_decay=1e-06,dropout=0.5,lr=0.0001:   PENDING                                                                                                                                                                               
RUNNING trials:                                                                                                                                                                                                                               
 - train_cifar10_0_decay=1e-06,dropout=0.25,lr=0.0001:  RUNNING                                                                                                                                                                               
                                                                                                                                                                                                                                              
Created LogSyncer for /home/nizam/ray_results/pbt_cifar10/train_cifar10_1_decay=1e-06,dropout=0.5,lr=0.0001_2018-08-15_10-47-42vpjvazid ->                                                                                                    
Epoch 1/1                                                                                                                                                                                                                                     
Epoch 1/1                                                                                                                                                                                                                                     
2018-08-15 10:47:44.581885: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA                                               
2018-08-15 10:47:44.587292: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA                                               
2018-08-15 10:47:44.696939: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero                   
2018-08-15 10:47:44.697002: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:897] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero                   
2018-08-15 10:47:44.697505: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1392] Found device 0 with properties:                                                                                                                          
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.8095                                                                                                                                                                         
pciBusID: 0000:01:00.0                                                                                                                                                                                                                        
totalMemory: 7.93GiB freeMemory: 6.74GiB                                                                                                                                                                                                      
2018-08-15 10:47:44.697516: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0                                                                                                                            
2018-08-15 10:47:44.697565: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1392] Found device 0 with properties:                                                                                                                          
name: GeForce GTX 1080 major: 6 minor: 1 memoryClockRate(GHz): 1.8095                                                                                                                                                                         
pciBusID: 0000:01:00.0                                                                                                                                                                                                                        
totalMemory: 7.93GiB freeMemory: 6.74GiB                                                                                                                                                                                                      
2018-08-15 10:47:44.697577: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1471] Adding visible gpu devices: 0                                                                                                                            
2018-08-15 10:47:44.846246: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:                                                                                           
2018-08-15 10:47:44.846270: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0                                                                                                                                                    
2018-08-15 10:47:44.846275: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N
2018-08-15 10:47:44.846385: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6448 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus
id: 0000:01:00.0, compute capability: 6.1)
2018-08-15 10:47:44.851194: I tensorflow/core/common_runtime/gpu/gpu_device.cc:952] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-08-15 10:47:44.851211: I tensorflow/core/common_runtime/gpu/gpu_device.cc:958]      0
2018-08-15 10:47:44.851215: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] 0:   N
2018-08-15 10:47:44.851290: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1084] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 110 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080, pci bus id: 0000:01:00.0, compute capability: 6.1)
2018-08-15 10:47:45.500122: E tensorflow/stream_executor/cuda/cuda_dnn.cc:332] could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
2018-08-15 10:47:45.501387: E tensorflow/stream_executor/cuda/cuda_dnn.cc:332] could not create cudnn handle: CUDNN_STATUS_INTERNAL_ERROR
^C/ray/src/local_scheduler/local_scheduler_client.cc:291 Check failed: static_cast<ray::protocol::MessageType>(type) == ray::protocol::MessageType::WaitReply
/home/nizam/anaconda3/envs/py36-tf-latest/lib/python3.6/site-packages/ray/core/src/local_scheduler/liblocal_scheduler_library_python.so(+0x18815)[0x7f3904fa3815]
/home/nizam/anaconda3/envs/py36-tf-latest/lib/python3.6/site-packages/ray/core/src/local_scheduler/liblocal_scheduler_library_python.so(_ZN3ray8internal7CerrLogD1Ev+0x8e)[0x7f3904face1e]
/home/nizam/anaconda3/envs/py36-tf-latest/lib/python3.6/site-packages/ray/core/src/local_scheduler/liblocal_scheduler_library_python.so(_Z20local_scheduler_waitP24LocalSchedulerConnectionRKSt6vectorIN3ray8UniqueIDESaIS3_EEilb+0x256)[0x7f3904fb2496]
/home/nizam/anaconda3/envs/py36-tf-latest/lib/python3.6/site-packages/ray/core/src/local_scheduler/liblocal_scheduler_library_python.so(+0x1a489)[0x7f3904fa5489]
python(_PyCFunction_FastCallDict+0x91)[0x55b149ad0921]
python(+0x19cdfc)[0x55b149b5ddfc]
python(_PyEval_EvalFrameDefault+0x2fa)[0x55b149b8294a]
python(+0x196206)[0x55b149b57206]
python(+0x1971cf)[0x55b149b581cf]
python(+0x19ced5)[0x55b149b5ded5]
python(_PyEval_EvalFrameDefault+0x2fa)[0x55b149b8294a]
python(+0x196f8b)[0x55b149b57f8b]
python(+0x19ced5)[0x55b149b5ded5]
python(_PyEval_EvalFrameDefault+0x2fa)[0x55b149b8294a]
python(+0x196f8b)[0x55b149b57f8b]
python(+0x19ced5)[0x55b149b5ded5]
python(_PyEval_EvalFrameDefault+0x2fa)[0x55b149b8294a]
python(+0x196206)[0x55b149b57206]
python(+0x1971cf)[0x55b149b581cf]
python(+0x19ced5)[0x55b149b5ded5]
python(_PyEval_EvalFrameDefault+0x2fa)[0x55b149b8294a]
python(PyEval_EvalCodeEx+0x329)[0x55b149b58cb9]
python(PyEval_EvalCode+0x1c)[0x55b149b59a4c]
python(+0x214c44)[0x55b149bd5c44]
python(PyRun_FileExFlags+0xa1)[0x55b149bd6041]
python(PyRun_SimpleFileExFlags+0x1c4)[0x55b149bd6244]
python(Py_Main+0x624)[0x55b149bd9d24]
python(main+0xee)[0x55b149aa175e]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xe7)[0x7f3972a31b97]
python(+0x1c847b)[0x55b149b8947b]
[1]    19747 abort (core dumped)  python pbt_tune_cifar10_with_keras.py
```

**Comment by ericl** [comment]
Fixed in x-ray backend
